### PR TITLE
JSON serialiser now calculates hash over schema

### PR DESCRIPTION
**Pull request by BernhardBln**

### CODE REVIEW

File: factcast-core/pom.xml [hunk 1]
```diff
@@ -1,64 +1,71 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (uweschaefer) [inline factcast-core/pom.xml:1]: what is the change here?

File: factcast-core/src/main/java/org/factcast/core/snap/SnapshotId.java [hunk 2]
```diff
@@ -29,4 +29,5 @@
     @NonNull
     UUID uuid; // semantically optional, might be constant if not needed
                // (SnapshotProjections)
+
```
Comment (uweschaefer) [inline factcast-core/src/main/java/org/factcast/core/snap/SnapshotId.java:32]: :grin:

File: factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java [hunk 3]
```diff
@@ -158,10 +177,21 @@ public static JsonNode toJsonNode(Map<String, Object> jsonAsMap) {
...
         return objectMapper.readerFor(type).readValue(bytes);
-
     }
 
     public static ObjectMapper mapper() {
         return objectMapper;
     }
+
+    @SneakyThrows
+    public static long calculateHash(Class<?> projectionClass) {
```
Comment (uweschaefer) [inline factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:187]: this uses the FCJson internal objectmapper, but we agreed that people must be able to swap the obejctmapper used for eventserialization. therefore i'd suggest to move this out of here and into the eventserializer ?

File: factcast-factus-bin-snapser/pom.xml [hunk 4]
```diff
@@ -1,46 +1,53 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (uweschaefer) [inline factcast-factus-bin-snapser/pom.xml:1]: what is the change here?

File: factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java [hunk 5]
```diff
@@ -158,10 +177,21 @@ public static JsonNode toJsonNode(Map<String, Object> jsonAsMap) {
...
+
+        JsonSchema jsonSchema = schemaGen.generateSchema(projectionClass);
```
Comment (uweschaefer) [inline factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:189]: also i guess, the schema should be cached (Map<Type,Schema>)

File: factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java [hunk 6]
```diff
@@ -158,10 +177,21 @@ public static JsonNode toJsonNode(Map<String, Object> jsonAsMap) {
...
+
+        String schema = writer
+                .writeValueAsString(jsonSchema);
+
+        return schemaModifier.apply(schema)
+                .hashCode();
```
Comment (uweschaefer) [inline factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:195]: is the hashcode even guaranteed to be consistent? i'd expect taking the md5 of the string

File: factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java (hunk 5 above)
Comment (uweschaefer) [reply factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:189]: or better Ma<Type,Long> directly to the md5

**Review by uweschaefer — CHANGES_REQUESTED** [review]
i'd suggest 

* MD5 instead of schema hash
* not recalculating the hash for a given class

File: factcast-core/pom.xml (hunk 1 above)
Comment (BernhardBln) [reply factcast-core/pom.xml:1]: No clue, probably from spotless plugin?

File: factcast-core/src/main/java/org/factcast/core/snap/SnapshotId.java (hunk 2 above)
Comment (BernhardBln) [reply factcast-core/src/main/java/org/factcast/core/snap/SnapshotId.java:32]: das sah aber auch aus :D

File: factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java (hunk 5 above)
Comment (BernhardBln) [reply factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:189]: Done in AbstractSnapshotRepository

File: factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java (hunk 6 above)
Comment (BernhardBln) [reply factcast-core/src/main/java/org/factcast/core/util/FactCastJson.java:195]: changed

File: factcast-core/pom.xml [hunk 7]
```diff
@@ -1,64 +1,71 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.11.2</version>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-jsonSchema</artifactId>
+            <version>2.11.2</version>
```
Comment (BernhardBln) [inline factcast-core/pom.xml:38]: How does it work with renovate-bot, would we remove the version here? Also above, in databind?

File: factcast-factus-bin-snapser/pom.xml (hunk 4 above)
Comment (BernhardBln) [reply factcast-factus-bin-snapser/pom.xml:1]: added json-schema dependency

File: factcast-core/pom.xml (hunk 7 above)
Comment (uweschaefer) [reply factcast-core/pom.xml:38]: i think there is a version defined in the properties. "jackson.version"?

**Comment by BernhardBln** [comment]
@uweschaefer I cleaned up dependency management for jackson a bit:

- introduced property with version in root pom
- used this property for all jackson dependencies in bom
- removed version tag in all other poms

**Comment by BernhardBln** [comment]
@uweschaefer what does that mean?

```
Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.4.1:check (default) on project factcast: 
     Execution default of goal com.diffplug.spotless:spotless-maven-plugin:2.4.1:check failed: 
          No such reference 'origin/master' -
```

https://github.com/factcast/factcast/pull/1050/checks?check_run_id=1175889328

**Comment by uweschaefer** [comment]
crap. spotless is configured to change only files diffing compared to master. the message implies, that the repo is not fully cloned, but rather just a particular branch is sparse checked out. will look into it. thx

**Review by uweschaefer — APPROVED** [review]

**Comment by uweschaefer** [comment]
thx @BernhardBln

**Pull request merged**